iothub_client/src/iothub_client_core_ll.c

```diff
@@ -2550,66 +2599,43 @@ IOTHUB_CLIENT_RESULT IoTHubClientCore_LL_SetDeviceMethodCallback(IOTHUB_CLIENT_C
     else
     {
         IOTHUB_CLIENT_CORE_LL_HANDLE_DATA* handleData = (IOTHUB_CLIENT_CORE_LL_HANDLE_DATA*)iotHubClientHandle;
-        if (deviceMethodCallback == NULL)
+        bool unsubscribing = (deviceMethodCallback == NULL);
+
+        if ((result = VerifyMethodCallbackType(handleData, CALLBACK_TYPE_SYNC, unsubscribing)) != IOTHUB_CLIENT_OK)
         {
-            if (handleData->methodCallback.type == CALLBACK_TYPE_NONE)
-            {
-                /* Codes_SRS_IOTHUBCLIENT_LL_10_029: [ If deviceMethodCallback is NULL and the client is not subscribed to receive method calls, IoTHubClientCore_LL_SetDeviceMethodCallback shall fail and return IOTHUB_CLIENT_ERROR. ] */
-                LogError("not currently set to accept or process incoming messages.");
-                result = IOTHUB_CLIENT_ERROR;
-            }
-            else if (handleData->methodCallback.type == CALLBACK_TYPE_ASYNC)
-            {
-                /* Codes_SRS_IOTHUBCLIENT_LL_10_028: [If the user has subscribed using IoTHubClientCore_LL_SetDeviceMethodCallback_Ex, IoTHubClientCore_LL_SetDeviceMethodCallback shall fail and return IOTHUB_CLIENT_ERROR. ] */
-                LogError("Invalid workflow sequence. Please unsubscribe using the IoTHubClientCore_LL_SetDeviceMethodCallback_Ex function.");
-                result = IOTHUB_CLIENT_ERROR;
-            }
-            else
-            {
-                /*Codes_SRS_IOTHUBCLIENT_LL_02_019: [If parameter messageCallback is NULL then IoTHubClientCore_LL_SetMessageCallback shall call the underlying layer's _Unsubscribe function and return IOTHUB_CLIENT_OK.] */
-                /*Codes_SRS_IOTHUBCLIENT_LL_12_018: [If deviceMethodCallback is NULL, then IoTHubClientCore_LL_SetDeviceMethodCallback shall call the underlying layer's IoTHubTransport_Unsubscribe_DeviceMethod function and return IOTHUB_CLIENT_OK. ] */
-                /*Codes_SRS_IOTHUBCLIENT_LL_12_022: [ Otherwise IoTHubClientCore_LL_SetDeviceMethodCallback shall succeed and return IOTHUB_CLIENT_OK. ]*/
-                handleData->IoTHubTransport_Unsubscribe_DeviceMethod(handleData->deviceHandle);
-                handleData->methodCallback.type = CALLBACK_TYPE_NONE;
-                handleData->methodCallback.callbackSync = NULL;
-                handleData->methodCallback.userContextCallback = NULL;
-                result = IOTHUB_CLIENT_OK;
-            }
+            ;
```

Comment:
??

Reply:
set result to error...or at least add a comment

Reply:
@ericwol-msft It's not showing up great on GitHub UX but I think you're referring to the lines below.  Correct?  If yes I can add a LogError so it's a bit more clear, though the result is set in the `if` block and the function called logs bunches but for consistency/clarity LogError here is fine.  Assuming this is what you meant right?

```c
        if ((result = VerifyMethodCallbackType(handleData, CALLBACK_TYPE_SYNC, unsubscribing)) != IOTHUB_CLIENT_OK)
        {
            ;
        }
```

Reply:
Adding LogError and will get updated out momentarily.